packages/store/src/PackedCounter.sol

```diff
@@ -1,127 +1,139 @@
+/**
+ * Instance functions for PackedCounter
+ */
+library PackedCounterInstance {
+  error PackedCounter_InvalidLength(uint256 length);
 
   /**
    * Decode the accumulated counter
    * (first 7 bytes of packed counter)
    */
   function total(PackedCounter packedCounter) internal pure returns (uint256) {
-    return uint256(uint56(bytes7(PackedCounter.unwrap(packedCounter))));
+    return uint56(uint256(PackedCounter.unwrap(packedCounter)));
   }
 
   /**
    * Decode the counter at the given index
    * (5 bytes per counter after the first 7 bytes)
    */
-  function atIndex(PackedCounter packedCounter, uint256 index) internal pure returns (uint256) {
-    uint256 offset = 7 + index * 5;
-    return uint256(uint40(Bytes.slice5(PackedCounter.unwrap(packedCounter), offset)));
+  function atIndex(PackedCounter packedCounter, uint8 index) internal pure returns (uint256) {
+    unchecked {
+      return uint40(uint256(PackedCounter.unwrap(packedCounter) >> (ACC_BYTES + VAL_BYTES * index)));
+    }
   }
 
   /**
    * Set a counter at the given index, return the new packed counter
    */
   function setAtIndex(
     PackedCounter packedCounter,
-    uint256 index,
+    uint8 index,
     uint256 newValueAtIndex
   ) internal pure returns (PackedCounter) {
-    bytes32 rawPackedCounter = PackedCounter.unwrap(packedCounter);
+    if (newValueAtIndex > MAX_VAL) {
+      revert PackedCounter_InvalidLength(newValueAtIndex);
+    }
+
+    uint256 rawPackedCounter = uint256(PackedCounter.unwrap(packedCounter));
 
     // Get current lengths (total and at index)
     uint256 accumulator = total(packedCounter);
-    uint256 currentValueAtIndex = atIndex(packedCounter, uint8(index));
+    uint256 currentValueAtIndex = atIndex(packedCounter, index);
 
     // Compute the difference and update the total value
-    if (newValueAtIndex >= currentValueAtIndex) {
-      accumulator += newValueAtIndex - currentValueAtIndex;
-    } else {
-      accumulator -= currentValueAtIndex - newValueAtIndex;
+    unchecked {
+      if (newValueAtIndex >= currentValueAtIndex) {
+        accumulator += newValueAtIndex - currentValueAtIndex;
+      } else {
+        accumulator -= currentValueAtIndex - newValueAtIndex;
+      }
     }
 
     // Set the new accumulated value and value at index
-    uint256 offset = 7 + index * 5; // (7 bytes total length, 5 bytes per dynamic schema)
-    rawPackedCounter = Bytes.setBytes7(rawPackedCounter, 0, bytes7(uint56(accumulator)));
-    rawPackedCounter = Bytes.setBytes5(rawPackedCounter, offset, bytes5(uint40(newValueAtIndex)));
+    // (7 bytes total length, 5 bytes per dynamic schema)
+    uint256 offset;
+    unchecked {
+      offset = ACC_BYTES + VAL_BYTES * index;
+    }
+    uint256 mask = uint256(type(uint40).max) << offset;
+    rawPackedCounter = (rawPackedCounter & ~uint256(type(uint56).max)) | accumulator;
+    rawPackedCounter = (rawPackedCounter & ~mask) | ((newValueAtIndex << offset) & mask);
```

Comment:
do we need to apply the `mask` in `((newValueAtIndex << offset) & mask)`? Since we already check that `newValueAtIndex` is < MAX_VAL at the top of this function